br: add restore sql binding steps (#14926)
pingcap/docs-cn: br/br-snapshot-manual.md

本文按备份恢复的场景介绍快照备份和恢复的命令行，包括：

> - [TiDB 快照备份与恢复命令行手册](#tidb-快照备份与恢复命令行手册)
>   - [备份集群快照](#备份集群快照)
>   - [备份 TiDB 集群指定库表的数据](#备份-tidb-集群指定库表的数据)
>     - [备份单个数据库的数据](#备份单个数据库的数据)
>     - [备份单张表的数据](#备份单张表的数据)
>     - [使用表库过滤功能备份多张表的数据](#使用表库过滤功能备份多张表的数据)
>   - [备份数据加密](#备份数据加密)
>   - [恢复快照备份数据](#恢复快照备份数据)
>   - [恢复备份数据中指定库表的数据](#恢复备份数据中指定库表的数据)
>     - [恢复单个数据库的数据](#恢复单个数据库的数据)
>     - [恢复单张表的数据](#恢复单张表的数据)
>     - [使用表库功能过滤恢复数据](#使用表库功能过滤恢复数据)
>     - [恢复系统表中存储的执行计划绑定信息](#恢复系统表中存储的执行计划绑定信息)
>   - [恢复加密的快照备份数据](#恢复加密的快照备份数据)

Reviewer: ```suggestion
- [备份集群快照](#备份集群快照)
- [备份 TiDB 集群指定库表的数据](#备份-tidb-集群指定库表的数据)
    - [备份单个数据库的数据](#备份单个数据库的数据)
    - [备份单张表的数据](#备份单张表的数据)
    - [使用表库过滤功能备份多张表的数据](#使用表库过滤功能备份多张表的数据)
- [备份数据加密](#备份数据加密)
- [恢复快照备份数据](#恢复快照备份数据)
- [恢复备份数据中指定库表的数据](#恢复备份数据中指定库表的数据)
    - [恢复单个数据库的数据](#恢复单个数据库的数据)
    - [恢复单张表的数据](#恢复单张表的数据)
    - [使用表库功能过滤恢复数据](#使用表库功能过滤恢复数据)
    - [恢复系统表中存储的执行计划绑定信息](#恢复系统表中存储的执行计划绑定信息)
- [恢复加密的快照备份数据](#恢复加密的快照备份数据)
```
“TiDB 快照备份与恢复命令行手册”是一级标题，不需要在这里列出

    --log-file restorefull.log
```

### 恢复系统表中存储的执行计划绑定信息

如果你需要恢复原集群的执行计划绑定信息，执行 `br restore full` 命令，设置 `--with-sys-table` 并通过 `--filter` 或 `-f` 指定需要恢复的系统表。

下面是恢复 `mysql.bind_info` 表的示例：

```shell
br restore full \
    --pd "${PD_IP}:2379" \
    --filter 'mysql.bind_info' \
    --with-sys-table \

>     --ratelimit 128 \

Reviewer 2: restore 的时候，感觉默认可以不用设置 ratelimit

    --storage "s3://${backup_collection_addr}/snapshot-${date}?access-key=${access-key}&secret-access-key=${secret-access-key}" \
    --log-file restore_system_table.log
```

恢复完成后，可以通过 [`SHOW GLOBAL BINDINGS`](/sql-statements/sql-statement-show-bindings.md) 检查执行计划绑定信息的恢复情况：

```sql
SHOW GLOBAL BINDINGS;
```

> 当前执行计划绑定信息在备份恢复后的动态加载仍在优化中（相关的 issue 为 [#46527](https://github.com/pingcap/tidb/issues/46527) 和 [#46528](https://github.com/pingcap/tidb/issues/46528)），你需要手动刷新执行计划绑定信息。

Reviewer 3: 要不要等这2个 issue 关了, 再合并呢?

Author: 这两个 issue 还没有确定排期修复，估计要等很长时间

    --log-file restorefull.log
```

### 恢复系统表中存储的执行计划绑定信息

> 如果你需要恢复原集群的执行计划绑定信息，执行 `br restore full` 命令，并用 `--with-sys-table` 指定说明需要恢复系统表，通过 `--filter` 或 `-f` 指定恢复 `mysql.bind_info` 表。

Reviewer: ```suggestion
如果你需要恢复原集群的执行计划绑定信息，执行 `br restore full` 命令，设置 `--with-sys-table` 并通过 `--filter` 或 `-f` 指定需要恢复的系统表。
```

下面是恢复 `mysql.bind_info` 表的示例：

```shell
br restore full \
    --pd "${PD_IP}:2379" \
    --filter 'mysql.bind_info' \
    --with-sys-table \
    --ratelimit 128 \
    --storage "s3://${backup_collection_addr}/snapshot-${date}?access-key=${access-key}&secret-access-key=${secret-access-key}" \
    --log-file restore_system_table.log
```

> 

Reviewer: ```suggestion

关于恢复系统表的更多信息，请参考[恢复 `mysql` 数据库下的表](/br/br-snapshot-guide.md#恢复-mysql-数据库下的表)。

```

Author: 不建议在这里链接到 [恢复 `mysql` 数据库下的表]，链接部分的文档还未经过 QA 验证

> 恢复完成后，可以通过 [`SHOW GLOBAL BINDINGS`](../sql-statements/sql-statement-show-bindings.md) 检查执行计划绑定信息的恢复情况。

Reviewer: ```suggestion
恢复完成后，可以通过 [`SHOW GLOBAL BINDINGS`](/sql-statements/sql-statement-show-bindings.md) 检查执行计划绑定信息的恢复情况：
```

```sql

> mysql> SHOW GLOBAL BINDINGS;

Reviewer: ```suggestion
SHOW GLOBAL BINDINGS;
```
为了方便用户直接在文档站中复制代码块，不建议添加 mysql>

```

> 当前执行计划绑定信息在备份恢复后的动态加载仍在优化中（相关的 issue 为 [#46527](https://github.com/pingcap/tidb/issues/46527) 和 [#46528](https://github.com/pingcap/tidb/issues/46528)，你需要手动刷新执行计划绑定信息。

Reviewer: ```suggestion
当前执行计划绑定信息在备份恢复后的动态加载仍在优化中（相关的 issue 为 [#46527](https://github.com/pingcap/tidb/issues/46527) 和 [#46528](https://github.com/pingcap/tidb/issues/46528)），你需要手动刷新执行计划绑定信息。
```

> 当前执行计划绑定信息在备份恢复后的动态加载仍在优化中 [#46527](https://github.com/pingcap/tidb/issues/46527), [#46528](https://github.com/pingcap/tidb/issues/46528)，需要手动刷新执行计划绑定信息。

Reviewer 4: "ADMIN RELOAD BINDINGS" 是不是可以先作为流程的一步，没有必要提 46528，这个可以后续作为一个增强。

> 当前执行计划绑定信息在备份恢复后的动态加载仍在优化中 [#46527](https://github.com/pingcap/tidb/issues/46527), [#46528](https://github.com/pingcap/tidb/issues/46528)，需要手动刷新执行计划绑定信息。

Reviewer: ```suggestion
当前执行计划绑定信息在备份恢复后的动态加载仍在优化中（相关的 issue 为 [#46527](https://github.com/pingcap/tidb/issues/46527) 和 [#46528](https://github.com/pingcap/tidb/issues/46528)，你需要手动刷新执行计划绑定信息。
```

```sql

> -- 确保 mysql.bind_info 表中 builtin_pseudo_sql_for_bind_lock 的记录仅 1 行，如果多于 1 行，需要手动删除
> mysql> select count(*) from mysql.bind_info where original_sql = 'builtin_pseudo_sql_for_bind_lock';
> 
> mysql> delete from bind_info where original_sql = 'builtin_pseudo_sql_for_bind_lock' limit 1;
> 
> -- 强制重新加载 Binding 信息
> mysql> ADMIN RELOAD BINDINGS;

Reviewer: ```suggestion
-- 确保 mysql.bind_info 表中 builtin_pseudo_sql_for_bind_lock 的记录仅 1 行，如果多于 1 行，需要手动删除
SELECT count(*) FROM mysql.bind_info WHERE original_sql = 'builtin_pseudo_sql_for_bind_lock';
DELETE FROM bind_info WHERE original_sql = 'builtin_pseudo_sql_for_bind_lock' LIMIT 1;

-- 强制重新加载绑定信息
ADMIN RELOAD BINDINGS;
```
